source/AttributeStore.h

```diff
@@ -0,0 +1,233 @@
+	// Gets the attribute for the specified category, or nullptr.
+	const Attribute *GetAttribute(const AttributeCategory category) const;
+	Attribute *GetAttribute(const AttributeCategory category);
+	const AttributeEffect *GetEffect(const AttributeAccess access) const;
+	AttributeEffect *GetEffect(const AttributeAccess access);
+	// Sets the value of the specified attribute. If the attribute is not present, it is added to this collection
+	// with this value.
+	void Set(const AttributeAccess attribute, double value);
+	template <class T>
+	void Set(const T &attribute, double value);
+	void Set(const char *attribute, double value);
+
+	// Checks whether there are any attributes stored here.
+	bool empty() const;
+
+	// Gets the minimum allowed value of the attribute.
+	double GetMinimum(const AttributeAccess attribute) const;
+	template <class T>
+	double GetMinimum(const T &attribute) const;
+	double GetMinimum(const char *attribute) const;
+
+	// Loads data from the data node. This function can be called multiple times on an instance.
+	void Load(const DataNode &node);
+	// Writes the attributes into the data writer.
+	void Save(DataWriter &writer) const;
+
+	// Calculates how effectively the action described by the category can be done.
+	// Some actions, like active cooling, have different performance under different conditions,
+	// while others are either actioned or not. Returns 0 if the action cannot be executed.
+	// Only supported for attribute categories that describe actions.
+	double GetEfficiency(const AttributeCategory category, const bool allowFractional) const;
+	// Applies the effects of the specified action to the target. If the efficiency is 0 or negative,
+	// the call is a no-op. Only supported for attribute categories that describe actions.
+	void ApplyEffects(const AttributeCategory category, const AttributeStore target, const double efficiency = 1.) const;
+
+	// Determine whether the given number of instances of the given attributes can
+	// be added to this instance. If not, return the std::maximum number that can be added.
+	int CanAdd(const AttributeStore &other, int count) const;
+	void Add(const AttributeStore &other, const int count);
+	// Adds attributes the specified number of times.
+	template <class A>
+	void Add(const A &attribute, const AttributeStore &other, const int count);
+	void Add(const char *attribute, const AttributeStore &other, const int count);
+	template <class A>
+	void Add(const A &attribute, const double amount);
+	void Add(const char *attribute, const double amount);
+
+	// Calls the given function on all attributes.
+	void ForEach(const std::function<void(const AnyAttribute &, double)> &function) const;
+
+private:
+	template <typename A>
+	int CanAdd(const A &attribute, const AttributeStore &other, int count) const;
+
+private:
+	const static std::map<std::string, double> MINIMUM_OVERRIDES;
+	Dictionary textAttributes;
+	std::map<AttributeCategory, Attribute> categorizedAttributes;
```

Comment:
The `categorizedAttributes` and the `textAttributes` variables could use some comments, since they seem quite critical to this PR. I guess the comments could be something like:
- categorizedAttributes: attributes that are relevant during flight, and that are optimized for fast and efficient access.
- textAttributes: attributes that are relevant in outfitters, but not used during flight.

Did I understand their functions correctly?

Reply:
Pretty much. There are some text attributes that are used during flight - crew count matters for how smoothly a ship is controlled, for instance. But for the most part, the engine relies on categorized attributes during flight, yes.

Reply:
Shouldn't they also be part of categorizedAttributes?

For example for kamikaze missiles (that thus are controlled by a crew member?
```
firing
	crew 1
```

Or for (plugin) weapons that kill crew remotely?
```
damage
	crew 2
```

Reply:
That certainly sounds interesting, but I think that would work better as part of a follow-up PR.

Reply:
Agreed to keep that for a follow-up PR. For now let's only add the comments around those variables.